x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java

```diff
@@ -75,12 +81,27 @@ public TransportEsqlQueryAction(
             bigArrays
         );
         this.settings = settings;
+        this.transportService = transportService;
     }
 
     @Override
     protected void doExecute(Task task, EsqlQueryRequest request, ActionListener<EsqlQueryResponse> listener) {
         // workaround for https://github.com/elastic/elasticsearch/issues/97916 - TODO remove this when we can
-        requestExecutor.execute(ActionRunnable.wrap(listener, l -> doExecuteForked(task, request, l)));
+        final ActionListener<EsqlQueryResponse> wrappedListener;
+        final TimeValue timeout = timeout(request);
+        if (timeout != null) {
+            final ThreadPool threadPool = transportService.getThreadPool();
+            final var executor = threadPool.executor(EsqlPlugin.ESQL_THREAD_POOL_NAME);
+            wrappedListener = ListenerTimeouts.wrapWithTimeout(threadPool, timeout, executor, listener, l -> {
+                logger.debug("cancelling ESQL task {} on timeout", task);
+                final TaskManager taskManager = transportService.getTaskManager();
+                taskManager.cancelTaskAndDescendants((CancellableTask) task, "timeout", false, ActionListener.noop());
```

Comment:
Even on a timeout we should wait for the child tasks to complete before responding.

Reply:
I agree; we should wait for all outstanding tasks before responding. I deliberately chose not to wait because of the specific use case we are trying to address. Here, we want users to see the timeout error from Elasticsearch rather than the Cloud proxy. If we wait for outstanding tasks, it's possible that it might exceed the proxy's timeout before we can respond. However, thinking about this more, I believe we should wait and reduce the default timeout instead. Having a parent task completed before its child tasks isn't good.

Reply:
FWIW I still think we should not try and undercut the proxy timeout by default. Our response is just an exception saying `ESQL query timed out after {}`, this is not practically any more useful than the proxy timeout message. Put differently, I expect most users who see this response will just increase the timeout in the request until they start to get the proxy timeout message instead.

Reply:
+1 to what you said, @DaveCTurner. I would like to hear from @nik9000 and @costin. If we all agree, then we should not introduce the timeout parameter until we really need it.

Reply:
I can see some value in a timeout _parameter_. Although generally it's be better to time out on the client side since server-side timeouts will always be a little unreliable and/or late, but I think there are some clients that make it harder than they should to do this. And server-side timeouts do save the cost of tearing down the connection and opening a new one at least. I just don't think we should have a server-side timeout by default.